src/js/transfers/bridged-nep21-to-erc20/index.js

```diff
@@ -6,56 +6,78 @@ import { borshifyOutcomeProof } from './borshify-proof'
 import { COMPLETE, FAILED, SUCCESS } from '../statuses'
 import * as storage from '../storage'
 
-// Call contract given by `nep21Address` to withdraw (burn) the given `amount`
-// of NEP21 tokens for window.nearUserAddress. The raw transaction info is
-// returned. This info is then used in `checkStatus` to derive the proof needed
-// for releasing `amount` tokens to window.ethUserAddress on Ethereum.
+// Immediately return to caller, which adds an ID to the provided data, stores
+// it in localStorage, and then calls `checkStatus` (below), which will then
+// call actual `withdraw` function.
 export async function initiate (nep21Address, amount) {
-  const bridgeToken = new NearContract(
-    window.nearConnection.account(),
+  return {
+    amount,
     nep21Address,
-    { changeMethods: ['withdraw'] }
-  )
+    status: INITIALIZED
+  }
+}
 
-  // TODO: can cause redirect to NEAR Wallet, rendering `tx` inaccessible!
-  const tx = await bridgeToken.withdraw(
-    {
-      amount: String(amount),
-      recipient: window.ethUserAddress.replace('0x', '')
-    },
-    new BN('3' + '0'.repeat(14)) // 10x current default from near-api-js
-  )
+async function withdraw ({ id, nep21Address, amount }) {
+  // Calling `bridgeToken.withdraw`, below, may cause a redirect to NEAR
+  // Wallet. Once back at this page, near-api-js will add transaction info to
+  // `completedTransactions` iterator.
+  //
+  // Given that the code here will only be processed when status is
+  // INITIALIZED, if we don't find this tx id in `completedTransactions`, then
+  // we know we need to start the withdrawal.
+  const withdrawing = window.nearConnection
+    .completedTransactions.remove(tx => tx.meta.id === id)
 
-  const receiptIds = tx.transaction_outcome.outcome.receipt_ids
+  if (!withdrawing) {
+    const bridgeToken = new NearContract(
+      window.nearConnection.account(),
+      nep21Address,
+      { changeMethods: ['withdraw'] }
+    )
 
-  if (receiptIds.length !== 1) {
-    throw new Error(
-      `Withdrawal expects only one receipt, got ${receiptIds.length}. Full withdrawal transaction: ${
-        JSON.stringify(tx)
-      }`
+    await bridgeToken.withdraw(
+      {
+        amount: String(amount),
+        recipient: window.ethUserAddress.replace('0x', '')
+      },
+      {
+        gas: new BN('3' + '0'.repeat(14)), // 10x current default from near-api-js
+        meta: { id }
+      }
     )
-  }
+  } else {
+    // transaction includes `failed: true` and an error message if it failed
+    if (withdrawing.failed) return withdrawing
 
-  const txReceiptId = receiptIds[0]
+    const receiptIds = withdrawing.transaction_outcome.outcome.receipt_ids
```

Comment:
Maybe `TxResult` should have different fields from RPC result. So you can use:
```
withdrawing = {
 result: {a:3},
 raw_result: '{"a": 3}'
 receipts: {123: {}},
 transaction: {sign...}
 outcomes: {123: {}},
 logs: [...],
...
```

Reply:
Sorry, I'm not yet familiar enough with the fields from the RPC result to understand the recommendation. Here's how I was picturing the `type` in `near-api-js`:

```ts`
type CompletedTransaction = FinalExecutionOutcome & {
    meta: { [key: string]: string | number };
    complete: true;
    failed: boolean;
    error?: string;
};
```

What modification would you recommend?

Reply:
I thought about parsing final execution outcome into something that can be queried easier, but the rest is the same.

Reply:
I agree that dealing with `FinalExecutionOutcome` is painful, but this also seems like an improvement that's out-of-band for the current task.